File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;

class Solution {
    private PriorityQueue<Integer> pq1 = new PriorityQueue<>();
    private PriorityQueue<Integer> pq2 = new PriorityQueue<>((i1, i2) -> i2 - i1);
    
    public int[] solution(String[] operations) {
        for (String operation : operations) {
            String[] split = operation.split(" ");
            String command = split[0];
            int value = Integer.parseInt(split[1]);
            if (command.equals("I")) {
                insert(value);
                continue;
            }
            if (value == -1) {
                deleteMin();
                continue;
            }
            if (value == 1) {
                deleteMax();
                continue;
            }
        }
        
        return isEmpty()
            ? new int[] {0, 0}
            : new int[] {pq2.peek(), pq1.peek()};
    }
    
    private void insert(int number) {
        pq1.add(number);
        pq2.add(number);
    }
    
    private void deleteMin() {
        if (isEmpty()) {
            return;
        }
        pq2.remove(pq1.poll());
    }
    
    private void deleteMax() {
        if (isEmpty()) {
            return;
        }
        pq1.remove(pq2.poll());
    }
    
    private boolean isEmpty() {
        return pq1.isEmpty() || pq2.isEmpty();
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.kt
import java.util.PriorityQueue

class Solution {
    private val minHeapQueue = PriorityQueue<Int>()
    private val maxHeapQueue = PriorityQueue<Int>(compareBy {-it})
    private val map = HashMap<Int, Int>()
        
    fun solution(operations: Array<String>): IntArray {
        
        
        operations.asSequence()
            .map {it.split(" ")}
            .forEach {
                val operation = it[0]
                val num = it[1].toInt()
                
                if (operation == "I") {
                    add(num)
                    return@forEach
                }
                
                if (num == 1) {
                    remove(maxHeapQueue)
                } else {
                    remove(minHeapQueue)
                }
            }
        
        return intArrayOf(element(maxHeapQueue), element(minHeapQueue))
    }
    
    fun add(num: Int) {
        minHeapQueue.add(num)
        maxHeapQueue.add(num)
        map[num] = (map[num] ?: 0) + 1
    }
    
    fun remove(queue: PriorityQueue<Int>) {
        cleanGargage(queue)
        
        if (queue.isEmpty()) {
            return
        }
        
        val num = queue.remove()
        
        map[num] = (map[num] ?: 1) - 1
        
        if (map[num] == 0) {
            map.remove(num)
        }
    }
    
    fun element(queue: PriorityQueue<Int>): Int {
        cleanGargage(queue)
        return when (queue.isEmpty()) {
            true -> 0
            false -> queue.element()
        }
    }
    
    fun cleanGargage(queue: PriorityQueue<Int>) {
        while (queue.isNotEmpty() && map[queue.element()] == null) {
            queue.remove()
        }
    }
}